### PR TITLE
feat(weave): Add Bedrock Guardrails support and Scorer

### PR DESCRIPTION
**Pull request by tcapelle**
This PR adds:
- Bedrock guardrails support thourhg the apply_guardrail API
- Improve readability of the paylods when using the invoke method
- Simple `BedrockGuardrailScorer` wrapper around the `apply_guardrail` API.

## Summary by CodeRabbit

- **New Features**
  - Integrated AWS-based guardrail functionality to evaluate and filter generated content.
  - Enhanced API processing for safer text generation.
- **Documentation**
  - Added a guide on using AWS guardrail features, including prerequisites and examples.
- **Tests**
  - Expanded testing to cover guardrail scenarios and error handling, including various edge cases.
- **Chores**
  - Updated dependencies to support additional integrations, including the addition of the `litellm` package.

### CODE REVIEW

File: docs/docs/guides/evaluation/guardrails_and_monitors.md [hunk 1]
```diff
@@ -178,6 +177,60 @@ async def generate_with_monitoring(prompt: str) -> str:
     return result
 ```
 
+## AWS Bedrock Guardrails
+
+The `BedrockGuardrailScorer` leverages AWS Bedrock's guardrail functionality to detect and filter content based on configured policies. It utilised the `apply_guardrail` API to apply the guardrail to the content.
+
+
+This scorer requires:
```
[the remark this answers is not in the document]
Comment (andrewtruong) [reply docs/docs/guides/evaluation/guardrails_and_monitors.md:185]: per your pyproject

File: docs/docs/guides/evaluation/guardrails_and_monitors.md
Comment (andrewtruong) [inline docs/docs/guides/evaluation/guardrails_and_monitors.md:1]: @J2-D2-3PO if you want to review docs here :)

File: weave/scorers/bedrock_guardrails.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+            )
+
+            # Check if the guardrail intervened
+            passed = response.get("action") != "GUARDRAIL_INTERVENED"
+
+            # Get the assessments
+            assessments = (
+                response.get("assessments", [{}])[0]
+                if response.get("assessments")
+                else {}
```
Comment (andrewtruong) [inline weave/scorers/bedrock_guardrails.py:87]: I think these are redundant

File: weave/scorers/bedrock_guardrails.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+                    "usage": response.get("usage", {}),
+                    "assessments": assessments,
+                },
+            )
+        except Exception as e:
+            return WeaveScorerResult(
+                passed=False,
+                metadata={
+                    "reason": f"Error applying Bedrock guardrail: {str(e)}",
+                    "error": str(e),
```
Comment (andrewtruong) [inline weave/scorers/bedrock_guardrails.py:108]: Why are you storing the error message twice?

File: weave/scorers/bedrock_guardrails.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    Note:
+        This scorer requires AWS Bedrock client to be properly configured with
+        appropriate permissions to access the guardrail API.
+    """
+
+    guardrail_id: str = Field(description="The identifier of the guardrail to use.")
+    guardrail_version: str = Field(
+        default="DRAFT", description="The version of the guardrail to use."
+    )
+    source: str = Field(
```
Comment (andrewtruong) [inline weave/scorers/bedrock_guardrails.py:29]: If these are really restricted, you should do `Literal["INPUT", "OUTPUT"]`

File: tests/integrations/bedrock/bedrock_test.py [hunk 5]
```diff
@@ -146,6 +210,18 @@ def mock_invoke_make_api_call(self, operation_name, kwarg):
     return orig(self, operation_name, kwarg)
 
 
+def mock_apply_guardrail_make_api_call(self, operation_name, kwarg):
```
Comment (andrewtruong) [inline tests/integrations/bedrock/bedrock_test.py:213]: do you mean `kwargs`?

File: tests/integrations/bedrock/bedrock_test.py [hunk 6]
```diff
@@ -300,3 +376,85 @@ def test_bedrock_invoke(client: weave.trace.weave_client.WeaveClient) -> None:
...
+            == MOCK_APPLY_GUARDRAIL_RESPONSE["outputs"][0]["text"]
+        )
+
+        # Check usage data
+        assert result.metadata["usage"]["inputTokens"] == 25
+        assert result.metadata["usage"]["outputTokens"] == 45
+        assert result.metadata["usage"]["totalTokens"] == 70
+
+    # Now verify that a trace was captured
+    calls = list(client.calls())
```
Comment (andrewtruong) [inline tests/integrations/bedrock/bedrock_test.py:426]: you don't need to call list here btw

File: weave/scorers/bedrock_guardrails.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        guardrail_version (str): The version of the guardrail to use.
+        source (str): The source of the content to evaluate, either 'INPUT' or 'OUTPUT'.
+
+    Note:
+        This scorer requires AWS Bedrock client to be properly configured with
+        appropriate permissions to access the guardrail API.
+    """
+
+    guardrail_id: str = Field(description="The identifier of the guardrail to use.")
+    guardrail_version: str = Field(
```
Comment (andrewtruong) [inline weave/scorers/bedrock_guardrails.py:26]: What is this?

File: docs/docs/guides/evaluation/guardrails_and_monitors.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (tcapelle) [reply docs/docs/guides/evaluation/guardrails_and_monitors.md:185]: removed

File: weave/integrations/bedrock/bedrock_sdk.py [hunk 8]
```diff
@@ -48,15 +51,53 @@ def bedrock_on_finish_invoke(
...
+        try:
+            # Try to parse as JSON for the logged copy
+            body_str = body_content.decode("utf-8")
+            outputs_copy["body"] = json.loads(body_str)
+        except (UnicodeDecodeError, json.JSONDecodeError):
+            # If we can't decode or parse, just use the raw content
+            outputs_copy["body"] = {"raw_content": str(body_content)}
+
+        # Create a BytesIO object from the content for the original response
+        outputs["body"] = io.BytesIO(body_content)
```
[the remark this answers is not in the document]
Comment (tcapelle) [reply weave/integrations/bedrock/bedrock_sdk.py:90]: This is weird boto3 logics, I do this so I can log the content without consuming it.

File: tests/integrations/bedrock/bedrock_test.py (hunk 5 above)
Comment (tcapelle) [reply tests/integrations/bedrock/bedrock_test.py:213]: and if it is only one 😍

File: weave/scorers/bedrock_guardrails.py (hunk 7 above)
Comment (tcapelle) [reply weave/scorers/bedrock_guardrails.py:26]: terrible boto3 api

File: tests/integrations/bedrock/bedrock_test.py (hunk 5 above)
Comment (andrewtruong) [reply tests/integrations/bedrock/bedrock_test.py:213]: but it's implicitly a dict here right?

File: docs/docs/guides/evaluation/guardrails_and_monitors.md (hunk 1 above)
Comment (J2-D2-3PO) [inline docs/docs/guides/evaluation/guardrails_and_monitors.md:185]:
```suggestion
The `BedrockGuardrailScorer` uses AWS Bedrock's guardrail feature to detect and filter content based on configured policies. It calls the `apply_guardrail` API to apply the guardrail to the content.

To use the `BedrockGuardrailScorer`, you need the following:
```

File: docs/docs/guides/evaluation/guardrails_and_monitors.md [hunk 9]
```diff
@@ -178,6 +177,60 @@ async def generate_with_monitoring(prompt: str) -> str:
...
+- An AWS account with Bedrock access
+- A configured guardrail in the AWS Bedrock console
+- The `boto3` Python package installed
+
+We actually create the bedrock client for you, so you don't need to create your own. If you need to specify a region pass the `bedrock_runtime_kwargs` parameter to the scorer.
+
+
+Note: You can find more information about how to create a guardrail in the [this notebook](https://github.com/aws-samples/amazon-bedrock-samples/blob/main/responsible_ai/bedrock-guardrails/guardrails-api.ipynb) from Bedrock.
```
Comment (J2-D2-3PO) [inline docs/docs/guides/evaluation/guardrails_and_monitors.md:193]:
```suggestion
- An AWS account with access to Bedrock
- A configured guardrail in the AWS Bedrock console
- The `boto3` Python package

:::tip
You don't need to create your own Bedrock client—Weave creates it for you.  To specify a region, pass the `bedrock_runtime_kwargs` parameter to the scorer.
:::

For more details on creating a guardrail, see the [Bedrock guardrails notebook](https://github.com/aws-samples/amazon-bedrock-samples/blob/main/responsible_ai/bedrock-guardrails/guardrails-api.ipynb).
```

File: docs/docs/guides/evaluation/guardrails_and_monitors.md [hunk 10]
```diff
@@ -178,6 +177,60 @@ async def generate_with_monitoring(prompt: str) -> str:
...
+guardrail_scorer = BedrockGuardrailScorer(
+    guardrail_id="your-guardrail-id",  # Replace with your guardrail ID
+    guardrail_version="DRAFT",         # Or use a specific version
+    source="INPUT",                   # Can be "INPUT" or "OUTPUT"
+    bedrock_runtime_kwargs={"region_name": "us-east-1"}  # AWS region
+)
+
+@weave.op
+def generate_text(prompt: str) -> str:
+    # Your text generation logic here
```
Comment (J2-D2-3PO) [inline docs/docs/guides/evaluation/guardrails_and_monitors.md:213]:
```suggestion
    # Add your text generation logic here
```

File: docs/docs/guides/evaluation/guardrails_and_monitors.md [hunk 11]
```diff
@@ -178,6 +177,60 @@ async def generate_with_monitoring(prompt: str) -> str:
...
+
+# Initialize Weave
+weave.init("my_app")
+
+# Create a guardrail scorer
+guardrail_scorer = BedrockGuardrailScorer(
+    guardrail_id="your-guardrail-id",  # Replace with your guardrail ID
+    guardrail_version="DRAFT",         # Or use a specific version
+    source="INPUT",                   # Can be "INPUT" or "OUTPUT"
+    bedrock_runtime_kwargs={"region_name": "us-east-1"}  # AWS region
```
Comment (J2-D2-3PO) [inline docs/docs/guides/evaluation/guardrails_and_monitors.md:208]:
```suggestion
    guardrail_id="your-guardrail-id",  # Replace "your-guardrail-id" with your guardrail ID
    guardrail_version="DRAFT",          # Use guardrail_version to use a specific guardrail version
    source="INPUT",                             # Can be "INPUT" or "OUTPUT"
    bedrock_runtime_kwargs={"region_name": "us-east-1"}  # AWS region
```

File: docs/docs/guides/evaluation/guardrails_and_monitors.md
Comment (J2-D2-3PO) [reply docs/docs/guides/evaluation/guardrails_and_monitors.md:1]: thanks @andrewtruong I reviewed and left some style suggestions

File: weave/scorers/bedrock_guardrails.py (hunk 7 above)
Comment (andrewtruong) [reply weave/scorers/bedrock_guardrails.py:26]: But what does it mean?  I don't understand what this setting does

File: file::memory:?cache=shared
Comment (andrewtruong) [inline file::memory:?cache=shared:1]: remove this file

File: tests/integrations/bedrock/bedrock_guardrail_test.py [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+            # Mock the apply_guardrail method
+            client.apply_guardrail = MagicMock()
+            return client
+
+    @pytest.fixture
+    def scorer(self, mock_bedrock_client):
+        """Create a BedrockGuardrailScorer instance with a mock client."""
+        scorer = BedrockGuardrailScorer(
+            guardrail_id="test-guardrail-id",
+            guardrail_version="DRAFT",
```
Comment (andrewtruong) [inline tests/integrations/bedrock/bedrock_guardrail_test.py:106]: I would call these `bedrock_guardrail_*` to disambiguate from weave keys.

File: tests/integrations/bedrock/bedrock_guardrail_test.py [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+            source="OUTPUT",
```
Comment (andrewtruong) [inline tests/integrations/bedrock/bedrock_guardrail_test.py:107]: this one too if it's specific to bedrock.  I'm not sure what it's for

File: tests/integrations/bedrock/bedrock_guardrail_test.py [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+
+        # Score the content and verify the client was called with guardrailVersion=2
+        scorer.score("Test content")
+        call_args = mock_bedrock_client.apply_guardrail.call_args[1]
+        assert call_args["guardrailVersion"] == "2"
+
+
+if __name__ == "__main__":
+    # This allows running the tests directly with pytest
+    pytest.main(["-xvs", __file__])
```
Comment (andrewtruong) [inline tests/integrations/bedrock/bedrock_guardrail_test.py:300]: what is this?  Remove

File: tests/integrations/bedrock/bedrock_test.py [hunk 15]
```diff
@@ -300,3 +378,85 @@ def test_bedrock_invoke(client: weave.trace.weave_client.WeaveClient) -> None:
...
+    guardrail_version = "DRAFT"
+    source = "OUTPUT"
+    bedrock_kwargs = {"region_name": "us-east-1"}
+
+    scorer = BedrockGuardrailScorer(
+        guardrail_id=guardrail_id,
+        guardrail_version=guardrail_version,
+        source=source,
+        bedrock_runtime_kwargs=bedrock_kwargs,
+    )
```
Comment (andrewtruong) [inline tests/integrations/bedrock/bedrock_test.py:399]: Why not just inline it?

**Review by andrewtruong — APPROVED** [review]
